src/main/kotlin/game/collections/Army.kt

```diff
@@ -1,36 +1,64 @@
 package game.collections
 
+import game.characters.Warlord
+import game.characters.Warrior
+import game.commands.Receiver
 import game.enums.WarriorType
 import game.factory.getWarrior
 import game.interfaces.BaseWarrior
+import game.interfaces.BaseWeapon
+
+class Army : Receiver() {
 
-class Army {
-    private val troops = mutableListOf<BaseWarrior>()
     private var currentWarrior: BaseWarrior? = null
+
+
     fun addUnits(quantity: Int, type: WarriorType) {
+
         repeat(quantity) {
             val warrior = getWarrior(type)
-            if (troops.isEmpty()) {
-                currentWarrior = warrior
-                troops.add(warrior)
+            if (warrior is Warlord && !this.containsWarlord()) {
+
+                warrior.warriorIfFront = currentWarrior
+                units.add(warrior)
+
             } else {
-                currentWarrior?.warriorBehind = warrior.also { currentWarrior = warrior }
-                troops.add(warrior)
+                if (units.isEmpty() && warrior !is Warlord) {
+                    currentWarrior = warrior
+                    units.add(warrior)
+                } else if (warrior !is Warlord) {
+                    currentWarrior?.warriorBehind = warrior.also { currentWarrior = warrior }
+                    units.add(warrior)
+                    warrior.warriorIfFront = units[units.indexOf(warrior) - 1]
```

Comment:
is that possible to have such case, when our warrior will have `index = 0`, in this case it should broke this program, isnt it ?